### PR TITLE
feat: add task manager

### PR DESCRIPTION
**Pull request by trestinlsd**
## Description
Add task manager to manage task   with exocore chain

----

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -4,6 +4,9 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/ExocoreNetwork/exocore/x/avs_task"
+	taskKeeper "github.com/ExocoreNetwork/exocore/x/avs_task/keeper"
```
Comment (mikebraver) [inline app/app.go]: Better rename to `avsTaskKeeper` and `avsTaskTypes` for avs indication

File: app/app.go [hunk 2]
```diff
@@ -658,6 +664,7 @@ func NewExocoreApp(
 	app.RewardKeeper = *rewardKeeper.NewKeeper(appCodec, keys[rewardTypes.StoreKey], app.AssetsKeeper)
 	app.ExoSlashKeeper = slashKeeper.NewKeeper(appCodec, keys[exoslashTypes.StoreKey], app.AssetsKeeper)
 	app.AVSManagerKeeper = *avsManagerKeeper.NewKeeper(appCodec, keys[avsManagerTypes.StoreKey], &app.OperatorKeeper, app.AssetsKeeper)
+	app.TaskKeeper = taskKeeper.NewKeeper(appCodec, keys[taskTypes.StoreKey], taskTypes.AvsKeeperMock{})
```
Comment (mikebraver) [inline app/app.go]: Please update the Mock to real avsKeeper

File: precompiles/avsTask/task.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	taskKeeper taskKeeper.Keeper
+	avsKeeper  tasktype.AvsKeeperMock
+}
+
+// NewPrecompile creates a new task Precompile instance as a
+// PrecompiledContract interface.
+func NewPrecompile(
+	authzKeeper authzkeeper.Keeper,
+	taskKeeper taskKeeper.Keeper,
+	avsKeeper tasktype.AvsKeeperMock,
```
Comment (mikebraver) [inline precompiles/avsTask/task.go]: Please change all mock to real implement type

File: precompiles/avsTask/task.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	if !contract.UseGas(cost) {
+		return nil, vm.ErrOutOfGas
+	}
+
+	return bz, nil
+}
+
+// IsTransaction checks if the given methodID corresponds to a transaction or query.
+//
+// Available task transactions are:
```
Comment (mikebraver) [inline precompiles/avsTask/task.go]: complete the comments: Available task transactions are: `MethodRegisterAVSTask `

File: proto/exocore/task/v1/tx.proto [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+import "amino/amino.proto";
+import "cosmos/msg/v1/msg.proto";
+import "cosmos_proto/cosmos.proto";
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/ExocoreNetwork/exocore/x/avs_task/types";
+// TaskContractInfo is the task info.
+message TaskContractInfo {
+    string TaskContractAddress = 1;
```
Comment (mikebraver) [inline proto/exocore/task/v1/tx.proto]: Fix all the proto-lint errors

File: x/avs_task/keeper/keeper.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+		cdc       codec.BinaryCodec
+		storeKey  storetypes.StoreKey
+		avsKeeper tasktype.AvsKeeperMock
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey storetypes.StoreKey,
+	avsKeeper tasktype.AvsKeeperMock,
```
Comment (mikebraver) [inline x/avs_task/keeper/keeper.go]: Please change the mock to real keeper too here

Comment (trestinlsd) [reply x/avs_task/keeper/keeper.go]: fixed

File: proto/exocore/task/v1/tx.proto (hunk 5 above)
Comment (trestinlsd) [reply proto/exocore/task/v1/tx.proto]: Fixed

File: precompiles/avsTask/task.go (hunk 3 above)
Comment (trestinlsd) [reply precompiles/avsTask/task.go]: Fixed

File: app/app.go (hunk 2 above)
Comment (trestinlsd) [reply app/app.go]: fixed

File: precompiles/avsTask/task.go (hunk 4 above)
Comment (trestinlsd) [reply precompiles/avsTask/task.go]: fixed

File: x/avstask/client/cli/tx.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+		RegisterAVSTask(),
+	)
+	return cmd
+}
+
+func RegisterAVSTask() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "Registertask params to taskManager module",
+		Short: "Registertask params to taskManager module",
+		Args:  cobra.ExactArgs(2),
```
Comment (TimmyExogenous) [inline x/avstask/client/cli/tx.go]: The number of arguments should be 3.

File: x/avstask/keeper/task.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+func (k *Keeper) GetAvsTaskInfo(ctx sdk.Context, addr string) (info *types.TaskContractInfo, err error) {
+	taskAccAddr, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "GetTaskInfo: error occurred when parse acc address from Bech32")
+	}
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixAVSTaskInfo)
+	// key := common.HexToAddress(incentive.Contract)
+	isExist := store.Has(taskAccAddr)
+	if !isExist {
+		return nil, errorsmod.Wrap(types.ErrNoKeyInTheStore, fmt.Sprintf("GetOperatorInfo: key is %suite", addr))
```
Comment (TimmyExogenous) [inline x/avstask/keeper/task.go]: What's the meaning of `%suite`, is it a mistake?

File: x/avstask/keeper/task.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+func (k *Keeper) GetOperatorPubKey(ctx sdk.Context, addr string) (pub []byte, err error) {
+	opAccAddr, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "GetTaskInfo: error occurred when parse acc address from Bech32")
+	}
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixOperatePub)
+	// key := common.HexToAddress(incentive.Contract)
+	isExist := store.Has(opAccAddr)
+	if !isExist {
+		return nil, errorsmod.Wrap(types.ErrNoKeyInTheStore, fmt.Sprintf("GetOperatorInfo: key is %suite", addr))
```
Comment (TimmyExogenous) [inline x/avstask/keeper/task.go]: Same comment as above.

File: x/evm/keeper/precompiles.go [hunk 10]
```diff
@@ -107,15 +108,24 @@ func AvailablePrecompiles(
 	if err != nil {
 		panic(fmt.Errorf("failed to load avsManager precompile: %w", err))
 	}
+	taskPrecompile, err := taskPrecompile.NewPrecompile(authzKeeper, taskKeeper, avsManagerKeeper)
+	if err != nil {
+		panic(fmt.Errorf("failed to load  reward precompile: %w", err))
+	}
+	blsPrecompile, err := blsPrecompile.NewPrecompile(6000)
```
Comment (TimmyExogenous) [inline x/evm/keeper/precompiles.go]: Please replace 6000 with a clearly named constant.

File: x/avstask/types/errors.go [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package types
+
+import (
+	errorsmod "cosmossdk.io/errors"
+)
+
+// x/avs_task module sentinel errors
+var (
+	ErrSample           = errorsmod.Register(ModuleName, 1100, "sample error")
+	ErrNoKeyInTheStore  = errorsmod.Register(ModuleName, 0, "there is not the key for in the store")
```
Comment (bwhour) [inline x/avstask/types/errors.go]: there is no such key in the store

File: x/avstask/keeper/task.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+	"github.com/ExocoreNetwork/exocore/x/avstask/types"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (k *Keeper) SetAvsTaskInfo(ctx sdk.Context, info *types.RegisterAVSTaskReq) (err error) {
+	taskAccAddr, err := sdk.AccAddressFromBech32(info.Task.TaskContractAddress)
+	if err != nil {
+		return errorsmod.Wrap(err, "SetTaskInfo: error occurred when parse acc address from Bech32")
```
Comment (bwhour) [inline x/avstask/keeper/task.go:16]: SetAvsTaskInfo,  and like "SetAvsTaskInfo GetAvsTaskInfo", this AVS could be  like other places all big letter

File: x/avstask/keeper/task.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	bz := k.cdc.MustMarshal(info)
+
+	store.Set(taskAccAddr, bz)
+	return nil
+}
+
+func (k *Keeper) GetAvsTaskInfo(ctx sdk.Context, addr string) (info *types.TaskContractInfo, err error) {
+	taskAccAddr, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "GetTaskInfo: error occurred when parse acc address from Bech32")
```
Comment (bwhour) [inline x/avstask/keeper/task.go]: GetAvsTaskInfo

File: x/avstask/keeper/task.go (hunk 8 above)
Comment (bwhour) [inline x/avstask/keeper/task.go]: GetAvsTaskInfo

File: x/avstask/keeper/task.go [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+func (k *Keeper) IsExistTask(ctx sdk.Context, addr sdk.AccAddress) bool {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixAVSTaskInfo)
+	return store.Has(addr)
+}
+
+func (k *Keeper) SetOperatorPubKey(ctx sdk.Context, addr string, pub []byte) (err error) {
+	opAccAddr, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return errorsmod.Wrap(err, "SetTaskInfo: error occurred when parse acc address from Bech32")
```
Comment (bwhour) [inline x/avstask/keeper/task.go]: SetOperatorPubKey

File: x/avstask/keeper/task.go [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixOperatePub)
+
+	store.Set(opAccAddr, pub)
+	return nil
+}
+
+func (k *Keeper) GetOperatorPubKey(ctx sdk.Context, addr string) (pub []byte, err error) {
+	opAccAddr, err := sdk.AccAddressFromBech32(addr)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "GetTaskInfo: error occurred when parse acc address from Bech32")
```
Comment (bwhour) [inline x/avstask/keeper/task.go]: GetOperatorPubKey

File: precompiles/avsTask/methods.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+	_ common.Address,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	// check the invalidation of caller contract
+	callerAddress, _ := bech32.EncodeFromBase256("exo", contract.CallerAddress.Bytes())
+	avs, _ := p.avsKeeper.GetAVSInfo(ctx, callerAddress)
+
+	if avs.GetInfo() == nil {
```
Comment (mikebraver) [inline precompiles/avsTask/methods.go:39]: Please check the return of this function, should be error here indicating the avs didn't exist
`func (k Keeper) GetAVSInfo(ctx sdk.Context, addr string) (*types.QueryAVSInfoResponse, error) `

File: precompiles/avsTask/task.go [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+	return bz, nil
+}
+
+// IsTransaction checks if the given methodID corresponds to a transaction or query.
+//
+// Available avstask transactions are:
+//   - RegisterAVSTask
+func (Precompile) IsTransaction(methodID string) bool {
+	switch methodID {
+	case MethodRegisterAVSTask:
```
Comment (mikebraver) [inline precompiles/avsTask/task.go]: Should add bls methods

File: precompiles/avsTask/parser.go [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+func (p Precompile) GetTaskParamsFromInputs(_ sdk.Context, args []interface{}) (*types.RegisterAVSTaskReq, error) {
+	if len(args) != 3 {
+		return nil, fmt.Errorf(cmn.ErrInvalidNumberOfArgs, 3, len(args))
+	}
+	taskParams := &types.RegisterAVSTaskReq{}
+	taskinfo := &types.TaskContractInfo{}
+
+	taskaddr, ok := args[0].(string)
+	if !ok {
```
Comment (mikebraver) [inline precompiles/avsTask/parser.go]: Also need to check if taskaddr == nil

File: precompiles/avsTask/parser.go [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 0, "string", taskaddr)
+	}
+	taskinfo.TaskContractAddress = taskaddr
+
+	taskName, ok := args[1].(string)
+	if !ok {
```
Comment (mikebraver) [inline precompiles/avsTask/parser.go]: same here

File: precompiles/avsTask/parser.go [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 1, "string", taskName)
+	}
+	taskinfo.Name = taskName
+
+	metainfo, ok := args[2].(string)
+	if !ok || metainfo == "" {
```
Comment (mikebraver) [inline precompiles/avsTask/parser.go:33]: same here

File: precompiles/bls/bls.go [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+		bz, err = p.AddTwoPubkeys(method, args)
+	default:
+		return nil, fmt.Errorf("invalid method")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// fmt.Printf("calling method: %s\n", method.Name)
```
Comment (mikebraver) [inline precompiles/bls/bls.go]: delete the useless comment

File: precompiles/bls/bls.go [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+	// fmt.Printf("precompile result: %x\n", bz)
+	return bz, nil
+}
```
Comment (mikebraver) [inline precompiles/bls/bls.go:106]: Where is `isTransaction()` function? please refer to above task.go

File: x/avstask/keeper/msg_server.go [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+	"github.com/ExocoreNetwork/exocore/x/avstask/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var _ types.MsgServer = &Keeper{}
+
+func (k *Keeper) RegisterAVSTask(ctx context.Context, req *types.RegisterAVSTaskReq) (*types.RegisterAVSTaskResponse, error) {
+	c := sdk.UnwrapSDKContext(ctx)
+	avs, _ := k.avsKeeper.GetAVSInfo(c, req.FromAddress)
+	if avs.GetInfo() == nil {
```
Comment (mikebraver) [inline x/avstask/keeper/msg_server.go:21]: same here

File: x/avstask/types/expected_keepers.go [hunk 24]
```diff
@@ -0,0 +1,18 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+)
+
+// AccountKeeper defines the expected account keeper used for simulations (noalias)
+type AccountKeeper interface {
```
Comment (mikebraver) [inline x/avstask/types/expected_keepers.go]: Why expected these two keepers? Seemed no use case in the pr

File: x/avstask/client/cli/tx.go (hunk 7 above)
Comment (trestinlsd) [reply x/avstask/client/cli/tx.go]: fixed

File: x/avstask/keeper/task.go (hunk 8 above)
Comment (trestinlsd) [reply x/avstask/keeper/task.go]: fixed

File: x/avstask/keeper/task.go (hunk 9 above)
Comment (trestinlsd) [reply x/avstask/keeper/task.go]: fixed

File: x/avstask/types/errors.go (hunk 11 above)
Comment (trestinlsd) [reply x/avstask/types/errors.go]: fixed

File: x/avstask/keeper/task.go (hunk 13 above)
Comment (trestinlsd) [reply x/avstask/keeper/task.go]: fixed

File: x/avstask/keeper/task.go (hunk 8 above)
Comment (trestinlsd) [reply x/avstask/keeper/task.go]: fixed

File: x/avstask/keeper/task.go (hunk 14 above)
Comment (trestinlsd) [reply x/avstask/keeper/task.go]: fixed

File: x/avstask/keeper/task.go (hunk 15 above)
Comment (trestinlsd) [reply x/avstask/keeper/task.go]: fixed

File: precompiles/avsTask/methods.go (hunk 16 above)
Comment (trestinlsd) [reply precompiles/avsTask/methods.go:39]: fixed

File: precompiles/avsTask/task.go (hunk 17 above)
Comment (trestinlsd) [reply precompiles/avsTask/task.go]: fixed

File: precompiles/bls/bls.go (hunk 21 above)
Comment (trestinlsd) [reply precompiles/bls/bls.go]: fixed

File: precompiles/bls/bls.go (hunk 22 above)
Comment (trestinlsd) [reply precompiles/bls/bls.go:106]: fixed

File: x/avstask/keeper/msg_server.go (hunk 23 above)
Comment (trestinlsd) [reply x/avstask/keeper/msg_server.go:21]: fixed

File: x/avstask/types/expected_keepers.go (hunk 24 above)
Comment (trestinlsd) [reply x/avstask/types/expected_keepers.go]: fixed

File: x/avstask/keeper/task.go (hunk 12 above)
Comment (trestinlsd) [reply x/avstask/keeper/task.go:16]: fixed

File: precompiles/avsTask/parser.go (hunk 18 above)
Comment (trestinlsd) [reply precompiles/avsTask/parser.go]: fixed

File: precompiles/avsTask/parser.go (hunk 19 above)
Comment (trestinlsd) [reply precompiles/avsTask/parser.go]: fixed

File: precompiles/avsTask/parser.go (hunk 20 above)
Comment (trestinlsd) [reply precompiles/avsTask/parser.go:33]: fixed

File: precompiles/avsTask/parser.go (hunk 18 above)
Comment (trestinlsd) [reply precompiles/avsTask/parser.go]: The parameters are all strings, and using=="" can determine whether they are empty

File: precompiles/avsTask/methods.go (hunk 16 above)
Comment (mikebraver) [reply precompiles/avsTask/methods.go:39]: what is `GetInfo()`? There's no such function

File: x/avstask/keeper/msg_server.go (hunk 23 above)
Comment (mikebraver) [reply x/avstask/keeper/msg_server.go:21]: Same here

File: precompiles/avsTask/methods.go (hunk 16 above)
Comment (trestinlsd) [reply precompiles/avsTask/methods.go:39]: This method exists

**Review by mikebraver — APPROVED** [review]

**Review by TimmyExogenous — APPROVED** [review]

**Review by bwhour — APPROVED** [review]

**Pull request merged**